Tasks/JavaToolInstaller/javatoolinstaller.ts

```diff
@@ -0,0 +1,122 @@
+        await getJava(versionSpec);
+    } catch (error) {
+        taskLib.error(error.message);
+        taskLib.setResult(taskLib.TaskResult.Failed, error.message);
+    } 
+}
+
+async function getJava(versionSpec: string) {
+    const fromAzure: boolean = ('AzureStorage' == taskLib.getInput('jdkSource', true));
+    const extractLocation: string = taskLib.getPathInput('jdkDestinationDirectory', true);
+    const cleanDestinationFolder: boolean = taskLib.getBoolInput('cleanDestinationFolder', false);
+    let compressedFileExtension: string;
+    let jdkDirectory: string;
+
+    toolLib.debug('Trying to get tool from local cache first');
+    const localVersions: string[] = toolLib.findLocalToolVersions('Java');
+    const version: string = toolLib.evaluateVersions(localVersions, versionSpec);
+
+     // Clean the destination folder before downloading and extracting?
+     if (cleanDestinationFolder && taskLib.exist(extractLocation) && taskLib.stats(extractLocation).isDirectory) {
+        console.log(taskLib.loc('CleanDestDir', extractLocation));
+
+        // delete the contents of the destination directory but leave the directory in place
+        fs.readdirSync(extractLocation)
+        .forEach((item: string) => {
+            const itemPath = path.join(extractLocation, item);
+            taskLib.rmRF(itemPath);
+        });
+    }
+
+    if (version) { //This version of Java JDK is already in the cache. Use it instead of downloading again.
+        console.log(taskLib.loc('Info_ResolvedToolFromCache', version));
+    } else if (fromAzure) { //Download JDK from an Azure blob storage location and extract.
+        console.log(taskLib.loc('RetrievingJdkFromAzure', version));
+        compressedFileExtension = getFileEnding(taskLib.getInput('azureCommonVirtualPath', true));
+    
+        const azureDownloader = new azureStorageDownloader.AzureStorageArtifactDownloader(taskLib.getInput('azureResourceManagerEndpoint', true), 
+            taskLib.getInput('azureStorageAccountName', true), taskLib.getInput('azureContainerName', true), taskLib.getInput('azureCommonVirtualPath', false));
+        await azureDownloader.downloadArtifacts(extractLocation, '*' + compressedFileExtension);
+        await sleepFor(250); //Wait for the file to be released before extracting it.
+
+        const extractSource = buildFilePath(extractLocation, compressedFileExtension);
+        jdkDirectory = new JavaFilesExtractor().unzipJavaDownload(extractSource, compressedFileExtension, extractLocation);
+    } else { //JDK is in a local directory. Extract to specified target directory.
+        console.log(taskLib.loc('RetrievingJdkFromLocalPath', version));
+        compressedFileExtension = getFileEnding(taskLib.getInput('jdkPath', true));
+        jdkDirectory = new JavaFilesExtractor().unzipJavaDownload(taskLib.getInput('jdkPath', true), compressedFileExtension, extractLocation);
+    }
+
+    unpackJars(jdkDirectory, path.join(jdkDirectory, 'bin'));
```

Comment:
I think the unpackJars call should be in the unzipJavaDownload method. If not, you are needlessly calling unpack even if the cache already contains this version of Java.